### PR TITLE
BREAKING CHANGE: Create a `Sender` or `SenderBuilder` from configuration or env var (#54)

### PR DESCRIPTION
**Pull request by amunra**
# Overview

The aim here is to support the new ILP config string format as documented in https://docs.google.com/document/d/1EQZjYrJO_Ll8RrSIZpbr15Ul3N-xAtTxvrunI-TpOR8 and now also supported by the Java client and soon other clients too.

# Changes

This introduces a new "configuration string" format that is being supported across all ILP clients.

The parser was already implemented as https://github.com/questdb/questdb-confstr-rs/.

The format is: `protocol::addr=host:port;key1=value1;key2=value2;...;`.

The protocols supported are:
* `tcp`
* `tcps` (with TLS)
* `http`
* `https` (with TLS)

The Rust `SenderBuilder`, and the C and C++ `opts` object methods have been aligned to the new configuration format.

For example in C++,

```
line_sender_builder opts{"localhost", 9009};
opts.auth("u", "t", "x", "y");
```

has been refactored to:

```
line_sender_builder opts{protocol::tcp, "localhost", 9009};
opts.username("u");
opts.token("t");
opts.token_x("x");
opts.token_y("y");
```

and is now equivalent to the conf string `tcp::addr=localhost:9009;username=u;token=t;token_x=x;token_y=y`

and can now support building directly without the `opts` object:

```
auto sender = line_sender::from_conf(
    "tcp::addr=localhost:9009;username=u;token=t;token_x=x;token_y=y");
```

The config string can also be read from an environment variable:

```
export QDB_CONF_STR="tcp::addr=localhost:9009;username=u;token=t;token_x=x;token_y=y"
```

```
auto sender = line_sender::from_env();
```

# HTTP

The config string can also be used to configure HTTP.

This is what creating a Rust client with HTTP+TLS looks like:

```
let sender = Sender::from_conf("https::addr=localhost:9000;username=admin;password=quest;");
```

For HTTP authentication you can also use a token:

```
let sender = Sender::from_conf("https::addr=localhost:9000;token=abc;");
```

# Progress

- [x] rename `SenderBuilder::new()` to `SenderBuilder::new_tcp()` and introduce a sibling `SenderBuilder::new_http()` ctor.
- [x] rename remaining parameters in the API to match config string.
- [x] change `-> Result<Self>` for all params to allow immediate validation failure.
- [x] test params can't override any set in config.
- [x] test tls can't be switched off for `tcps` and `https`
- [x] consider if the tcp-only `net_interface` should be part of the doc - ensure it can't be used for HTTP (I've got an open request with the `ureq` crate).
- [x] add `Sender::from_conf` and `Sender::from_env` constructors.
- [x] perform full validation logic in `SenderBuilder::connect`.
- [x] consider renaming `SenderBuilder::connect()` to `SenderBuilder::build()` -- since it would be misleading for HTTP where no connection takes place.
- [x] update the C api tracking the renames and new `err_out` params.
- [x] update the wrapper C++ api.
- [x] update the stub test python API, i.e. `system_test/questdb_line_sender.py`.
- [x] cross-code review Java and Rust implementation for any undesired behaviour discrepancies. Eg. consider if `grace_timeout` and `min_throughput` are still the valid names.
- [x] add integration tests for `from_conf` and `from_env` for both `SenderBuilder` and `Sender` APIs.
- [x] review and update C and C++ docs.
- [x] figure out why CI breaks with the latest questdb `master` code: Possibly `HEAD` is no longer supported for the `/` path on HTTP with the latest questdb changes in master.
- [x] default user agent for C and C++.
- [x] write new `from_conf` example for each of Rust / C / C++ and register them to the manifest file for cloud and docs.

### CODE REVIEW

**Comment by amunra** [comment]
Nightly has warnings:
```
warning: the item `itoa` is imported redundantly
   --> src/ingress/mod.rs:188:5
    |
188 | use itoa;
    |     ^^^^ the item `itoa` is already defined here
    |
    = note: `#[warn(unused_imports)]` on by default

warning: the item `TryFrom` is imported redundantly
   --> src/ingress/mod.rs:190:32
    |
190 | use std::convert::{Infallible, TryFrom, TryInto};
    |                                ^^^^^^^
   --> /rustc/0ecbd0605770f45c9151715e66ba2b3cae367fcb/library/std/src/prelude/mod.rs:129:13
    |
    = note: the item `TryFrom` is already defined here

warning: the item `TryInto` is imported redundantly
   --> src/ingress/mod.rs:190:41
    |
190 | use std::convert::{Infallible, TryFrom, TryInto};
    |                                         ^^^^^^^
   --> /rustc/0ecbd0605770f45c9151715e66ba2b3cae367fcb/library/std/src/prelude/mod.rs:129:13
    |
    = note: the item `TryInto` is already defined here

warning: the item `ToString` is imported redundantly
   --> src/ingress/mod.rs:196:5
    |
196 | use std::string::ToString;
    |     ^^^^^^^^^^^^^^^^^^^^^
   --> /rustc/0ecbd0605770f45c9151715e66ba2b3cae367fcb/library/std/src/prelude/mod.rs:125:13
    |
    = note: the item `ToString` is already defined here
warning: `questdb-rs` (lib) generated 4 warnings
  |     -------- the item `Result` is already imported here
2 | use crate::{error::Result, ErrorCode};
  |             ^^^^^^^^^^^^^

```

**Comment by amunra** [comment]
Windows build test assertion discrepancies:
```
ingress::tests::tcps_tls_roots_file_missing

thread 'ingress::tests::tcps_tls_roots_file_missing' panicked at src\ingress\tests.rs:452:5:
assertion `left == right` failed
test ingress::tests::tcps_tls_roots_file_missing ... FAILED
  left: "Could not open root certificate file from path \"/some/invalid/path/cacerts.pem\": The system cannot find the path specified. (os error 3)"
test ingress::tests::tcps_tls_roots_file ... ok
 right: "Could not open root certificate file from path \"/some/invalid/path/cacerts.pem\": No such file or directory (os error 2)"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

File: examples/concat.c [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#include "concat.h"
+
+/** Concatenate a list of nul-terminated strings. Pass an extra NULL arg to terminate. */
+char* concat_(const char* first, ...) {
+    va_list args;
+    size_t totalLength = strlen(first) + 1;
```
Comment (amunra) [inline examples/concat.c:6]: `totalLength` -> `total_length`

File: examples/line_sender_c_example.c [hunk 2]
```diff
@@ -2,29 +2,28 @@
...
-
-    line_sender_utf8 host_utf8 = { 0, NULL };
-    if (!line_sender_utf8_init(&host_utf8, strlen(host), host, &err))
-        goto on_error;
-
-    line_sender_utf8 port_utf8 = { 0, NULL };
-    if (!line_sender_utf8_init(&port_utf8, strlen(port), port, &err))
+    char* conf_str = concat("tcp::addr=", host, ":", port, ";");
+    if (!conf_str) {
+        fprintf(stderr, "Could not concatenate connection string.\n");
```
Comment (amunra) [inline examples/line_sender_c_example.c:14]: connection string -> configuration string

File: examples/line_sender_c_example_auth.c [hunk 3]
```diff
@@ -2,45 +2,33 @@
...
-
-    line_sender_utf8 host_utf8 = { 0, NULL };
-    if (!line_sender_utf8_init(&host_utf8, strlen(host), host, &err))
-        goto on_error;
-
-    line_sender_utf8 port_utf8 = { 0, NULL };
-    if (!line_sender_utf8_init(&port_utf8, strlen(port), port, &err))
+    char* conf_str = concat(
+        "tcp::addr=", host, ":", port, ";",
+        "username=testUser1;",
```
Comment (amunra) [inline examples/line_sender_c_example_auth.c:14]: remove redundant comma after `user1;"`.

File: .bumpversion.cfg [hunk 4]
```diff
@@ -26,3 +26,7 @@ replace = questdb-rs/{new_version}/
 [bumpversion:file:questdb-rs-ffi/Cargo.toml]
 search = version = "{current_version}"
 replace = version = "{new_version}"
+
+[bumpversion:file:include/questdb/line_sender.hpp]
```
Comment (amunra) [inline .bumpversion.cfg:30]: Fix to `include/questdb/ingress/line_sender.hpp`

Comment (amunra) [reply .bumpversion.cfg:30]: Done.

File: doc/CPP.md [hunk 5]
```diff
@@ -27,17 +27,15 @@
 
 ...
 
-// Automatically connects on object construction.
-questdb::ingress::line_sender sender{
-    "localhost",  // QuestDB hostname
-    9009};        // QuestDB port
+auto sender = questdb::ingress::line_sender::from_conf(
+    "http::addr=localhost:1;");
```
[the remark this answers is not in the document]
Comment (amunra) [reply doc/CPP.md:31]: Done.

File: examples/concat.c (hunk 1 above)
Comment (amunra) [reply examples/concat.c:6]: Done.

File: examples/line_sender_c_example.c (hunk 2 above)
Comment (amunra) [reply examples/line_sender_c_example.c:14]: Done (along with other similar instances).

File: examples/line_sender_c_example_auth.c (hunk 3 above)
Comment (amunra) [reply examples/line_sender_c_example_auth.c:14]: Done (and in other places too).

File: questdb-rs-ffi/src/lib.rs [hunk 6]
```diff
@@ -57,13 +57,76 @@ macro_rules! bubble_err_to_c {
...
+    // We need to extract a `T` from a `*mut T` and then replace it with
+    // another `T` in situ.
+    let dest = &mut (*opts).0;
+    let forced_builder = ptr::read(&(*opts).0);
+    let new_builder = match forced_builder.tls_roots(path) {
+        Ok(builder) => builder,
+        Err(err) => {
+            let err_ptr = Box::into_raw(Box::new(line_sender_error(err)));
+            *err_out = err_ptr;
+
```
[the remark this answers is not in the document]
Comment (amunra) [reply questdb-rs-ffi/src/lib.rs:74]: Done.

File: questdb-rs-ffi/src/lib.rs [hunk 7]
```diff
@@ -992,21 +925,326 @@ pub unsafe extern "C" fn line_sender_buffer_at_now(
...
 }
 
+/// Accumulates parameters for creating a line sender connection.
+pub struct line_sender_opts(SenderBuilder);
+
+/// Create a new `ops` instance from configuration string.
+/// The format of the string is: "tcp::addr=host:port;key=value;...;"
+/// Alongside "tcp" you can also specify "tcps", "http", and "https".
+/// The accepted set of keys and values is the same as for the opt's API.
+/// E.g. "tcp::addr=host:port;username=alice;password=secret;tls_ca=os_roots;"
```
Comment (amunra) [inline questdb-rs-ffi/src/lib.rs:935]: tcp:: -> http::

Comment (amunra) [reply questdb-rs-ffi/src/lib.rs:935]: Done.

**Review by mtopolnik — APPROVED** [review]

**Pull request merged**